optimise Lazy value generation and calculation

mkeskells opened the pull request:
to reduce memory and lock contention in multi-threaded jvm phase

lrytz commented:
Test failures:
```
[error] Test scala.tools.nsc.backend.jvm.BTypesTest.lazyLockForceTest failed: org.junit.ComparisonFailure: expected:<1-[2:hi-3:hi]:hi> but was:<1-[3:hi:hi-2]:hi>, took 0.003 sec
[error]     at scala.tools.nsc.backend.jvm.BTypesTest.lazyLockForceTest(BTypesTest.scala:91)
```

```
[error] Test scala.tools.nsc.backend.jvm.BTypesTest.lazyNoLockForceTest failed: org.junit.ComparisonFailure: expected:<1-[2:hi-3:hi]:hi> but was:<1-[3:hi:hi-2]:hi>, took 0.001 sec
[error]     at scala.tools.nsc.backend.jvm.BTypesTest.lazyNoLockForceTest(BTypesTest.scala:102)
```

mkeskells commented:
Yes – fixing and enhacing the test now

 

From: Lukas Rytz [mailto:notifications@github.com] 
Sent: 06 November 2017 21:34
To: scala/scala <scala@noreply.github.com>
Cc: mkeskells <mike.skells@talk21.com>; Author <author@noreply.github.com>
Subject: Re: [scala/scala] optimise Lazy value generation and calculation (#6162)

 

Test failures:

[error] Test scala.tools.nsc.backend.jvm.BTypesTest.lazyLockForceTest failed: org.junit.ComparisonFailure: expected:<1-[2:hi-3:hi]:hi> but was:<1-[3:hi:hi-2]:hi>, took 0.003 sec
[error]     at scala.tools.nsc.backend.jvm.BTypesTest.lazyLockForceTest(BTypesTest.scala:91)
[error] Test scala.tools.nsc.backend.jvm.BTypesTest.lazyNoLockForceTest failed: org.junit.ComparisonFailure: expected:<1-[2:hi-3:hi]:hi> but was:<1-[3:hi:hi-2]:hi>, took 0.001 sec
[error]     at scala.tools.nsc.backend.jvm.BTypesTest.lazyNoLockForceTest(BTypesTest.scala:102)

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/scala/scala/pull/6162#issuecomment-342294647> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AVPl5bhNwyIGRuE3l_-2Vago6r9nPk_Hks5sz3tYgaJpZM4QSkcH> .  <https://github.com/notifications/beacon/AVPl5cRXqWoAXWkVPN_OejYBiiJAA_B-ks5sz3tYgaJpZM4QSkcH.gif>

lrytz commented:
Thanks - I was just pointing it out because it's a bit hard to find in the logs

lrytz commented on src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala:
```diff
@@ -891,52 +891,113 @@ abstract class BTypes {
+      * A lazy value that synchronizes on the `frontendLock`, and supports accumulating actions
+      * to be executed when it's forced.
+      */
+    private final class LazyWithLock[T <: AnyRef](t: () => T) extends AbstractLazy[T](t) {
```
In the previous implementation the init function (`t` here) was set to `null` after being used. Maybe we should keep that to release object references?

mkeskells replied:
done

lrytz commented on src/compiler/scala/tools/nsc/backend/jvm/BTypesFromSymbols.scala:
```diff
@@ -422,12 +422,12 @@ abstract class BTypesFromSymbols[G <: Global](val global: G) extends BTypes {
     val nestedClasses = {
       val ph = phase
-      Lazy(enteringPhase(ph)(nestedClassSymbolsNoJavaModuleClasses.map(classBTypeFromSymbol)))
+      Lazy.withLock(enteringPhase(ph)(nestedClassSymbolsNoJavaModuleClasses.map(classBTypeFromSymbol)))
```
I wonder if we can identify certain `Lazy`s that we know will be forced, depending on the callsite where the `ClassBType` / `ClassInfo` is created from (code generation in particular)? Maybe you could experiment a bit (track from where the `ClassBType`s are created and see which ones will be completed)?

mkeskells replied:
I did have a brief try with making it more eager. I agree that if we know its going to be needed it better to do it more eagerly though.

I didn't spend much time on that though, but didnt see any obvious pattern

mkeskells replied:
in my testing of the 3 call sites that use the lock about 70% of them were expanded.

Another thought that I had was that the background threads could expand the non symbol Lazy values more eagerly if we know some that we will need ( the withoutLock values). This doesn't occur without optimisation as I see it and that isn't my focus really

mkeskells replied:
@lrytz I went back and had another look. Some improvements added and I think that there are more to come.Not sure if they are part of this PR though.Will discuss via email

lrytz commented on src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala:
```diff
@@ -891,52 +891,113 @@ abstract class BTypes {
+  abstract sealed class Lazy[T] {
+    /** get the result of the lazy value, calculation the result and performing the additional actions if the value
+      * is not already known.
+      * @return the
```
Let's keep the doc comments concise; the main comment is enough, no need for the annotations (also in the rest of the PR). Also, the comment here is a bit hard to read.

mkeskells replied:
done

lrytz commented on src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala:
```diff
@@ -891,52 +891,113 @@ abstract class BTypes {
-      else frontendSynch {
+      override def toString = if (value == null) "<?>" else value.toString
+
+      private var postForce: List[T => Unit] = Nil
```
The previous implementation had the advantage of not requiring a field for those actions - could we still use that approach now?

mkeskells replied:
I don't think that was a true saving. You had a inner class that wrapped the action via a dynamic class and it also extended the lock duration

If we are really concerned with these field counts we have save much more by making these top level classes and not using cake pattern everywhere

mkeskells commented:
I have fixed and added some test cases, clarified the execution order of accumulated functions (slightly different to pre-existing), and reduced the comments

lrytz commented on src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala:
```diff
@@ -891,52 +891,105 @@ abstract class BTypes {
+          result = init(t)
+          t = null
+          this.synchronized {
+            postForce foreach (_ (result))
```
the syntax here could be more explicit `postForce.foreach(_.apply(result))`

mkeskells replied:
I did consider that.... Have changed to be more explicit

I would not expect to see any material change to the performance from this PR, until we get a parallel genBCode

Maybe a slight memory allocation reduction, but I don't think that your benchmarks include that measure

lrytz reviewed (commented):
Benchmarks are scheduled, results should [come up here](https://scala-ci.typesafe.com/grafana/dashboard/db/pr-validation?var-scalaSha=38628d11fc0c67b552e9c640a7f544dd7563235c%7C6afe450f8e22ebe36fe04bf7b92f1a6f78f18ccf%7C28f9105763936c2da7ffe10ebd79561b35ac4f6a&var-source=All&var-bench=HotScalacBenchmark.compile&var-host=scalabench@scalabench@&from=1504445910328&to=1510322201039)

lrytz commented:
LGTM, but could you adjust the commit messages so that the first line is not too long (72, https://github.com/scala/scala/blob/2.13.x/CONTRIBUTING.md#clean-commits-clean-history)?

mkeskells committed:
optimise Lazy value generation and calculation
Reduce memory and lock contention in multi-threaded jvm phase
Single values for common eager values
Predict if expansion is certain to be required, and eagerly generate
in that case to reduce lock contention in parallel GenBCode compiles

mkeskells commented:
commit messages adjusted

jbgi commented on src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala:
```diff
@@ -891,52 +891,114 @@ abstract class BTypes {
+      def onForce(f: T => Unit): Unit = {
         if (value != null) f(value)
+        else this.synchronized {
+          if (value != null) f(value)
```
> assign to local var to avoid volatile reads and associated memory barriers

not done above.

jbgi commented on src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala:
```diff
@@ -891,52 +891,114 @@ abstract class BTypes {
+
+      def init(t: () => T): T = frontendSynch {
+        if (value == null) value = t()
+        value
```
> assign to local var to avoid volatile reads and associated memory barriers

not done above.
